### ISSUE TITLE
Add validated sensor model

### ISSUE THREAD

**Issue by cwwalter** [LSSTDESC/imSim#7]
The following features need to be added/validated to the GalSim module before we can use it for DC2.

-  [x] Leveraging the interface in GalSim-developers/GalSim#799 (mid-layer API for drawImage) 
-  [x] Validate the brighter-fatter code.  Being done by @craiglagegit and Kristen (no Github?) 
-  [x] Implement adding per-sensor orientation tree rings (#59).
-  [x] The model does saturation, and blooming.  We need to add bleed trails (in @jchiang87 readout code in #6?) 
- [x]  Add cosmic rays #76 
- [x] Performance tests #86

**Comment by cwwalter** [comment]
This code now exists in the master of GalSim.  However, it is still being validated.

**Comment by cwwalter** [comment]
I have updated this issue with a checkbox list, to add the features that need to be added/validated in the sensor model by DC2.

**Comment by cwwalter** [comment]
@janewman-pitt-edu points out that in regards to cosmic rays" "Ultimately it will affect the false positive rate for supernovae significantly (as some pictures effectively get half depth in a spatially correlated way) so it's be nice to have,".  

In my original discussion with @SimonKrughoff last year about this, DM wasn't ready to do this subtraction to remove them from the two snaps yet. Jeff says we should skip it for DC2 if that is still the case, but I thought I should recheck what the status of the DM code is for CR removal now to see if it is something we should plan on waiting until DC3 for.

**Comment by cwwalter** [comment]
Letting @rbiswas4 and @reneehlozek know about this issue and possibly get their feedback about CR removal etc.

**Comment by cwwalter** [comment]
Adding some information here from @jchiang87 from an email thread discussing effects in the amplifiers for future reference.

"We have measured non-linearity in the raft-level testing at the ~few
tenths of a percent level.   My code doesn't model that, but it could
easily be modified to do so.   Same applies to adding saturation
effects."

**Comment by cwwalter** [comment]
@rbiswas4 and @reneehlozek and @	janewman_pitt_edu you all requested or asked about cosmic rays in DC2.  

Cosmic rays will be added, this work is already almost finished by @jchiang87.   I updated the list above.

For others:

Saturation in the sensor alone is included in  @craiglagegit's model but we still need to add any extra saturation associated with the amplifier readout.  I wanted to record/remind ourselves the now closed #29 has some comments from @dkirkby on properly implementing and validating the saturation when there is a lot of charge in the buried channel.  We might want to recheck that with the current silicon model.

We also need to add bleed trail in the readout step but could use guidance or a model on the best way to model this.

**Comment by craiglagegit** [comment]
Chris and Jim, I just wanted to correct the above and day that saturation is not currently included in the GalSim silicon model that I  implemented.  It could be,  but Mike Jarvis and I  concluded it was better to add it at the end in the detector read out code. CraigOn Nov 18, 2017 7:56 PM, Chris Walter <notifications@github.com> wrote:@rbiswas4 and @reneehlozek and @	janewman_pitt_edu you all requested or asked about cosmic rays in DC2.
Cosmic rays will be added, this work is already almost finished by @jchiang87.   I updated the list above.
For others:
Saturation in the sensor alone is included in  @craiglagegit's model but we still need to add any extra saturation associated with the amplifier readout.  I wanted to record/remind ourselves the now closed #29 has some comments from @dkirkby on properly implementing and validating the saturation when there is a lot of charge in the buried channel.  We might want to recheck that with the current silicon model.
We also need to add bleed trail in the readout step but could use guidance or a model on the best way to model this.

**Comment by cwwalter** [comment]
> Chris and Jim, I just wanted to correct the above and day that saturation is not currently included in the GalSim silicon model that I  implemented.  It could be,  but Mike Jarvis and I  concluded it was better to add it at the end in the detector read out code. Craig

OK, I think we (I) should be more careful about language.  I was using saturation the way we usually use it for photomultiplier tubes.  Here I mean not just a hard cut off but any non linearity as we begin approach full well.

So, I think there are two non-linearity effects: one is in the silicon device itself. I believe you showed us that was now working well in SSim meeting and even responded correctly to applied voltage correct?  This is what I was referring too.

Then, there are non-linearities associated with the amplifiers and readout system and finally there is a the effect of bleeding over the column when the depleted well is really full which could really give a hard cuttoff.  These two effects should be in the readout code.

That is what I meant, do I have any misunderstandings?

**Comment by cwwalter** [comment]
If #86 closes I think we can close this too.  The validation part deserves its own effort related to the wiki etc.
